extensions/iceberg/src/main/java/io/deephaven/iceberg/util/SchemaSpec.java

```diff
@@ -0,0 +1,140 @@
+/**
+ * A specification for extracting the schema from a table.
+ */
+interface SchemaSpec {
+
+    /**
+     * Returns the schema for the given table based on this SchemaSpec.
+     */
+    Schema getSchema(Table table);
+
+    // Static factory methods for creating SchemaSpec instances
+    static SchemaSpec initial() {
+        return new InitialSchemaSpec();
+    }
+
+    static SchemaSpec fromSchemaId(final int id) {
+        return new IdSchemaSpec(id);
+    }
+
+    static SchemaSpec fromSchema(final Schema schema) {
+        return new DirectSchemaSpec(schema);
+    }
+
+    static SchemaSpec fromSnapshotId(final int snapshotId) {
+        return new SnapshotIdSchemaSpec(snapshotId);
+    }
+
+    static SchemaSpec latest() {
+        return new LatestSchemaSpec();
+    }
+
+    // --------------------------------------------------------------------------------------------------
+
+    // Implementations of SchemaSpec
+    class InitialSchemaSpec implements SchemaSpec {
+        @Override
+        public Schema getSchema(Table table) {
+            return getInitialSchema(table);
+        }
+    }
+
+    class IdSchemaSpec implements SchemaSpec {
+        private final int schemaId;
+
+        IdSchemaSpec(final int schemaId) {
+            this.schemaId = schemaId;
+        }
+
+        @Override
+        public Schema getSchema(final Table table) {
+            return getSchemaForId(table, schemaId);
+        }
+    }
+
+    class DirectSchemaSpec implements SchemaSpec {
+        private final Schema schema;
+
+        DirectSchemaSpec(final Schema schema) {
+            this.schema = schema;
+        }
+
+        @Override
+        public Schema getSchema(final Table table) {
+            return schema;
+        }
+    }
+
+    class SnapshotIdSchemaSpec implements SchemaSpec {
+        private final int snapshotId;
+
+        SnapshotIdSchemaSpec(final int snapshotId) {
+            this.snapshotId = snapshotId;
+        }
+
+        @Override
+        public Schema getSchema(Table table) {
+            return getSchemaForSnapshotId(table, snapshotId);
+        }
+    }
+
+    class LatestSchemaSpec implements SchemaSpec {
+        @Override
+        public Schema getSchema(Table table) {
+            return getLatestSchema(table);
+        }
+    }
+
+    // --------------------------------------------------------------------------------------------------
+
+    // Methods for extracting the schema from the table
+    private static Schema getInitialSchema(final Table table) {
+        final Iterable<Snapshot> snapshots = table.snapshots();
+        final Iterator<Snapshot> iterator = snapshots.iterator();
+
+        if (iterator.hasNext()) {
+            final Snapshot initialSnapshot = iterator.next();
+            final Integer initialSchemaId = initialSnapshot.schemaId();
+            if (initialSchemaId != null) {
+                getSchemaForId(table, initialSchemaId);
+            }
+        }
+        return table.schema();
```

Comment:
I was hoping to read the specs and find some guarantee about the "original" schema of the table, but it had no such guarantees afaict. The Java implementation uses `org.apache.iceberg.TableMetadata#INITIAL_SCHEMA_ID`, but again, there is no guarantee of that.

I was even trying to verify that schemas were never removed from a table metadata, but I couldn't even find that guarantee. Looking at the java implementation, I see that it never removes schemas, but I can't find that guarantee in the spec.

As such, I think we should remove initial; the alternative is to treat initial as `getSchemaForId(table, 0)` which would be "ok" for some tables, but I can't guarantee for all.

Reply:
Makes sense, I will remove this and use the latest schema by default.